### PR TITLE
[Core] Convert IntegrationEventsCallbacks To Class

### PR DESCRIPTION
**Pull request by VenfiOranai**
# Description

What - Converted IntegrationEventsCallbacks from typed dict to dataclass 

Why - Makes it significantly more traceable within the codebase

How -

## Release (integrations / ocean core)

If this PR changes an integration or `port_ocean/`, add a release intent file **or** manually bump `pyproject.toml` and `CHANGELOG.md` (manual takes priority):

- Integration: `integrations/<name>/.ocean-release/<unique-name>.yaml`
- Core: `.ocean-release/core/<unique-name>.yaml`

```yaml
bump: patch
changelog-type: bugfix
changelog: Describe the change
```

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

---

> [!NOTE]
> **Medium Risk**
> Touches core event registration and resync/start dispatch. Behavior should be equivalent, but any external code that treated `event_strategy` as a dict will break.
> 
> **Overview**
> Converts `IntegrationEventsCallbacks` from a `TypedDict` to a dataclass so event listeners are accessed as attributes (`event_strategy.resync`) instead of dict keys.
> 
> `EventsMixin` now constructs the dataclass (still using `defaultdict(list)` for kind maps). Call sites in `BaseIntegration` and `SyncRawMixin` use attribute access; dynamic `"resync"` / `"incremental"` lookup goes through `getattr`. `resync_complete` hooks always run (the old `"resync_complete" in ...` guard is gone). Tests are updated to match.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 22b9899f7c0d50b02be9060962891db9aeb5a9f8. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: port_ocean/core/integrations/mixins/sync_raw.py [hunk 1]
```diff
@@ -94,10 +93,8 @@ async def _get_resource_raw_results(
             else self.available_resync_kinds
         )
 
-        if not is_resource_supported(
-            resource_config.kind, self.event_strategy[strategy_key]
-        ):
-            return unsupported_kind_response(resource_config.kind, available_kinds)
+        if resource_config.kind not in available_kinds and None not in available_kinds:
+            return unsupported_kind_response(resource_config.kind, cast(list[str], available_kinds))
```
Comment (raskinmaya) [inline port_ocean/core/integrations/mixins/sync_raw.py]: Removing `is_resource_supported` changes behavior. Checking `available_*_kinds` passes on empty lists and mutates `defaultdict`s, whereas the old check ensured non-empty handlers. Please restore `mapping.get(kind) or mapping.get(None)` to keep original semantics

Comment (raskinmaya) [inline port_ocean/core/integrations/mixins/sync_raw.py]: Another one here:
`available_*_kinds` includes `None`, so using `cast(list[str])` just masks the type issue and prints `Available kinds: [None]` on errors. Instead of using `cast`, either update `KindNotImplementedException` to accept `list[str | None]` or filter/format `None` when raising it

File: port_ocean/core/integrations/mixins/sync_raw.py [hunk 2]
```diff
@@ -111,9 +108,10 @@ async def _get_resource_raw_results(
     def _collect_resync_functions(
         self, resource_config: ResourceConfig, strategy_key: str = "resync"
     ) -> list[Callable[[str], Awaitable[RAW_RESULT]]]:
+        event_mapping = getattr(self.event_strategy, strategy_key)
```
Comment (raskinmaya) [inline port_ocean/core/integrations/mixins/sync_raw.py]: `getattr` strips type checking and risks silent bugs on invalid strings. Since `strategy_key` is strictly `"resync" | "incremental"`, swap `getattr` for an explicit if/else or a typed method like `get_mapping(key: Literal["resync", "incremental"])`

File: port_ocean/core/integrations/mixins/sync_raw.py [hunk 3]
```diff
@@ -94,10 +93,8 @@ async def _get_resource_raw_results(
             else self.available_resync_kinds
         )
 
-        if not is_resource_supported(
-            resource_config.kind, self.event_strategy[strategy_key]
-        ):
-            return unsupported_kind_response(resource_config.kind, available_kinds)
+        if resource_config.kind not in available_kinds and None not in available_kinds:
```
Comment (raskinmaya) [inline port_ocean/core/integrations/mixins/sync_raw.py]: The old path used O(1) dict lookups via `is_resource_supported`. The new path materializes `list(keys())` on every resync, then two O(n) in scans. Integrations like ADO register many kinds - this runs per resource. A `mapping.get(kind) or mapping.get(None)` check is cheaper and matches old semantics

File: port_ocean/core/integrations/mixins/sync_raw.py [hunk 4]
```diff
@@ -1003,13 +1001,12 @@ async def _resync_reconciliation(
...
-                logger.info("Executing resync_complete hooks")
+            logger.info("Executing resync_complete hooks")
 
-                for resync_complete_fn in self.event_strategy["resync_complete"]:
-                    await resync_complete_fn()
+            for resync_complete_fn in self.event_strategy.resync_complete:
+                await resync_complete_fn()
 
-                logger.info("Finished executing resync_complete hooks")
+            logger.info("Finished executing resync_complete hooks")
```
Comment (raskinmaya) [inline port_ocean/core/integrations/mixins/sync_raw.py:1014]: DSP skips `_resync_reconciliation`, so this doesn't apply there, but for ocean-core processing mode, we now always log `Executing/Finished executing resync_complete hooks` even when `resync_complete` is empty if I understand correctly. Most integrations don't register this hook, so that's two info lines per successful full resync. Not too critical, but why not guard with `if self.event_strategy.resync_complete`?

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 1 above)
Comment (raskinmaya) [inline port_ocean/core/integrations/mixins/sync_raw.py]: Please add test coverage for the missing edge cases around `is_resource_supported` since this is the riskiest part of the PR:
- Specific kind registered
- Catch-all only
- Unregistered kind
- Both catch-all and specific kind registered

File: port_ocean/core/ocean_types.py [hunk 5]
```diff
@@ -53,9 +53,14 @@ class ETLPhase:
...
-class IntegrationEventsCallbacks(TypedDict):
-    start: list[START_EVENT_LISTENER]
-    resync: dict[str | None, list[RESYNC_EVENT_LISTENER]]
-    resync_start: list[BEFORE_RESYNC_EVENT_LISTENER]
-    resync_complete: list[AFTER_RESYNC_EVENT_LISTENER]
-    incremental: dict[str | None, list[INCREMENTAL_EVENT_LISTENER]]
+@dataclass
+class IntegrationEventsCallbacks:
+    start: list[START_EVENT_LISTENER] = field(default_factory=list)
+    resync: dict[str | None, list[RESYNC_EVENT_LISTENER]] = field(
```
Comment (raskinmaya) [inline port_ocean/core/ocean_types.py:59]: Not too critical, but `default_factory=lambda: defaultdict(list)` is duplicated. Extracting it to a named function or helper (e.g. `_default_listener_map`) would clean this up and eliminate the repetition. Non-blocking

File: .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml
Comment (raskinmaya) [inline .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml:1]: Directly mutating `integration.event_strategy["resync"][kind]` will break after this PR is in. Since `@ocean.on_resync()` is the official API, this is low-risk, but it can still be a breaking change for private (customer-owned) integrations. WDYT about bumping a minor version instead of a patch?

**Review by raskinmaya — CHANGES_REQUESTED** [review]

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 1 above)
Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py]: Not super sure I get the behavior change you're talking about here. If `available_kinds` is empty then it'll enter the if due to the first half of the condition (just like it used), and I'm not really seeing what part here can potentially mutate a defaultdict

Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py]: But we have the `None not in available_kinds` check in the if, so we know for sure it's only strings. Only reason I added the casting is since the accursed IDE considers that a runtime check so it didn't recognize it.

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 3 above)
Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py]: Not sure the extra cognitive load is worth it. Going off a quick check, even if we have 1000 kinds (which I assume we don't), it takes less than a millisecond.
I can convert to a set if you think it'd be preferable, but not sure there's much of a point.

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 4 above)
Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py:1014]: Cursor added that guard originally, but that actually changes behavior. Notice that the if we had there checked that the key exists, not that there's a value in it. The key of course always existed since it was added into the dict on init, so the logs always fired no matter what. I can add the if, but that's a change I'm not sure I want attributed to this PR (in case somebody's debugging and doesn't get why logs they used to have suddenly disappeared)

File: port_ocean/core/ocean_types.py (hunk 5 above)
Comment (VenfiOranai) [reply port_ocean/core/ocean_types.py:59]: Feels a bit like overkill for a lambda that just calls a defaultdict

File: .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml
Comment (VenfiOranai) [reply .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml:1]: I agree that it's a breaking change, thing is we should technically bump majors for those, so not sure there's too much of a diff between minor and patch. I'll add a mention in the changelog for sure, but from what I've seen we treat anything internal/bug as a patch, and features as a minor. Since this is fully internal, I went with patch.
I can change it if you think it's preferable, whatever you think

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 2 above)
Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py]: Nice catch, converted to if/else

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 1 above)
Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py]: Nice catch, added

File: .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml
Comment (raskinmaya) [reply .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml:1]: I agree, but at least not a patch. Let's bump a minor version

File: port_ocean/core/integrations/mixins/sync_raw.py (hunk 3 above)
Comment (raskinmaya) [reply port_ocean/core/integrations/mixins/sync_raw.py]: Customer mapping size drives the multiplier, not handler count - and `mapping.get(kind) or mapping.get(None)` avoids that on the success path. I'd rather we change that

Comment (VenfiOranai) [reply port_ocean/core/integrations/mixins/sync_raw.py]: Ah, okie, done (plus a bit more to avoid replicated ifs)

File: .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml
Comment (VenfiOranai) [reply .ocean-release/core/task_tk9lzk_convert_integrationeventscallbacks_to_class.yaml:1]: Done

**Review by raskinmaya — APPROVED** [review]
